File: workers/common/rbm.ts
export const packBitmap1bit = (rows: number[][], width: number, height: number): Uint8Array => {
  const stride = Math.ceil(width / 8);
  const out = new Uint8Array(stride * height);
  for (let y = 0; y < height; y++) {
    const row = rows[y];
    let byte = 0;
    let bit = 7;
    let idx = y * stride;
    for (let x = 0; x < width; x++) {
      if (row && row[x]) byte |= (1 << bit);
      if (--bit < 0) {
        out[idx++] = byte;
        byte = 0;
        bit = 7;
      }
    }
    if (bit !== 7) out[idx] = byte;
  }
  return out;
};

export const encodeRBM = (bits: Uint8Array, w: number, h: number, seq: number, durMs = 0): Uint8Array => {
  const buf = new Uint8Array(16 + bits.length);
  buf[0] = 0x52; buf[1] = 0x42; // 'RB'
  buf[2] = 1; // version
  buf[3] = 0; // flags
  buf[4] = (w >> 8) & 0xff; buf[5] = w & 0xff;
  buf[6] = (h >> 8) & 0xff; buf[7] = h & 0xff;
  buf[8] = (seq >>> 24) & 0xff; buf[9] = (seq >>> 16) & 0xff; buf[10] = (seq >>> 8) & 0xff; buf[11] = seq & 0xff;
  buf[12] = (durMs >> 8) & 0xff; buf[13] = durMs & 0xff;
  buf[14] = 0; buf[15] = 0;
  buf.set(bits, 16);
  return buf;
};

File: workers/common/rbm.py
from __future__ import annotations

import math
import struct
from typing import Iterable, List


def pack_bitmap_1bit(rows: Iterable[Iterable[int]], width: int, height: int) -> bytes:
    """Pack a 2D array of 0/1 ints into MSB-first row-major bytes.

    rows: iterable of rows, each an iterable of 0/1 values, length=height
    width, height: dimensions of the logical canvas
    """
    stride = (width + 7) // 8
    out = bytearray(height * stride)
    y = 0
    for row in rows:
        x = 0
        byte_idx = y * stride
        current = 0
        bitpos = 7
        for v in row:
            if v:
                current |= 1 << bitpos
            bitpos -= 1
            x += 1
            if bitpos < 0:
                out[byte_idx] = current
                byte_idx += 1
                current = 0
                bitpos = 7
            if x >= width:
                break
        if bitpos != 7:
            out[byte_idx] = current
        y += 1
        if y >= height:
            break
    return bytes(out)


def encode_rbm(
    frame_bits: bytes,
    width: int,
    height: int,
    seq: int = 0,
    frame_duration_ms: int = 0,
    flags: int = 0,
) -> bytes:
    """Encode RBM header + payload (see protocol/rbm_spec.md)."""
    magic = b"RB"
    version = 1
    header = struct.pack(
        ">2sBBHHIHH",
        magic,
        version,
        flags,
        width,
        height,
        seq & 0xFFFFFFFF,
        frame_duration_ms & 0xFFFF,
        0,  # reserved
    )
    return header + frame_bits
